plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/CloudLibraries.java

```diff
@@ -65,28 +72,47 @@
   }
   
   /**
-   * @return the library with the specified ID, or null if not found
+   * Returns the library with the specified ID, or null if not found.
    */
   public static Library getLibrary(String id) {
-    // todo this method is only called by LibraryClasspathContainerResolverService and tests.
-    // Perhaps here or in LibraryClasspathContainerResolverService is where we should 
-    // resolve the transitive dependencies
     return libraries.get(id);
   }
+
+  private static final LoadingCache<IJavaProject, Library> masterLibraries =
+      CacheBuilder.newBuilder().build(new CacheLoader<IJavaProject, Library>() {
```

Comment:
Still trying to understand what the cache does. If we read/write JSON and can figure out which jars the master container has, why do we need a cache? (Certainly adding new APIs is an infrequent event, and even if it were frequent, such an event can hardly be a performance bottleneck.)

Reply:
There are multiple master libraries, one per project. The cache is just a map from JavaProject to master library with some stronger thread-safety guarantees and simpler initialization than we would get from a HashMap.

Reply:
Wee could do that, yes. However it means we add the initialization logic, including the check as to whether a master library for the project already exists, in all three places. 
And we'd still need to store at least  a map of JavaProjects --> Master Libraries, and all three locations would have to update this map. It seemed simpler to me to have it in a single method call.